File: app/src/main/java/com/example/david/lists/view/authentication/IAuthContract.java
package com.example.david.lists.view.authentication;

public interface IAuthContract {
    interface View {
        void signIn(int requestCode);

        void signOut();

        void displayMessage(String message);

        void setResult(IAuthContract.AuthResult result);

        void setResultFailed(String reason);

        void finishView();
    }

    interface Logic {
        void onStart(IAuthContract.AuthGoal authGoal);

        void signInSuccessful();

        void signInCancelled();

        void signInFailed(int errorCode);

        void signOutSucceeded();

        void signOutFailed(Exception e);
    }

    interface ViewModel {
        int getRequestCode();

        String getMsgSignInSucceed();

        String getMsgSignInCanceled();

        String getMsgSignOutSucceed();

        String getMsgSignInError(int errorCode);

        String getMsgSignOutFailed();

        String getMsgSignInWhenNotAnon();
    }


    enum AuthResult {
        AUTH_SUCCESS,
        AUTH_FAILED,
        // the user cancelled.
        AUTH_CANCELLED
    }

    // Need better name.
    enum AuthGoal {
        SIGN_IN,
        SIGN_OUT,
        // Authenticate an anonymous user.
        AUTH_ANON
    }
}

File: app/src/main/java/com/example/david/lists/view/authentication/IAuthContract.kt
package com.example.david.lists.view.authentication

interface IAuthContract {
    interface View {
        fun signIn(requestCode: Int)

        fun signOut()

        fun displayEmailSentMessage(email: String)

        fun hideEmailSentMessage()

        fun displayMessage(message: String)

        fun openMainView()

        fun finishView()
    }

    interface Logic {
        fun onStart(signOut: Boolean = false)

        fun signInSuccessful()

        fun signInCancelled()

        fun signInFailed(errorCode: Int)

        fun signOutSucceeded()

        fun signOutFailed(e: Exception)

        fun verifyEmailButtonClicked()
    }

    interface ViewModel {
        val signInRequestCode: Int

        var emailVerificationSent: Boolean

        val msgSignInSucceed: String

        val msgSignInCanceled: String

        val msgSignInError: String

        fun getMsgSignInError(errorCode: Int): String

        val msgEmailNotVerified: String

        val msgSignOutSucceed: String

        val msgSignOutFailed: String
    }
}
